### PR TITLE
[Modules] Update SA/blobservice - deleteRetentionPolicy fix and addition

### PR DESCRIPTION
**Pull request by MariusStorhaug**
# Description

- Rename the mapped value for `deleteRetentionPolicyEnabled` passed from `storage-account/main.bicep` to `blob-services/main.bicep`. Today this is 'deleteRetentionPolicy'. As there is no documentation stating that the expected property of the passed `blobServices` object in storage-account should be `deleteRetentionPolicy`, this PR suggests aligning this with the Azure Resource Reference. 
- Add missing property. Path: `properties.deleteRetentionPolicy.allowPermanentDelete`.
- Updated module `README.md` for storage-account.

## Pipeline references

| Pipeline |
| - |
| [Storage - StorageAccounts](https://github.com/MariusStorhaug/ResourceModules/actions/workflows/ms.storage.storageaccounts.yml) |

# Type of Change

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Update to documentation

# Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (readme)
- [x] I did format my code

### CODE REVIEW

**Comment by MariusStorhaug** [comment]
@microsoft-github-policy-service agree

**Review by AlexanderSehr — CHANGES_REQUESTED** [review]
Jojojo Mr. @MariusStorhaug , 
thanks a lot for the Pull Request. Just one ask - please run `bicep build` for the Bicep template so that also the ARM JSON template is updated.

Also, do you have a test environment available? If not, we can merge the PR into a branch and try to test it from our end (and add the test to the test cases).

**Comment by MariusStorhaug** [comment]
Hey @AlexanderSehr. I can run the json generation. Seem there is missing some doc that you want both files. Would you want the metadata in the `main.json` however? The following is added now:

image

As for the test env, I currently do not have a functioning one. I can set it up but will be a bit later this week.

**Comment by AlexanderSehr** [comment]
> Hey @AlexanderSehr. I can run the json generation. Seems the wiki is missing some doc on the main.json file (or I skimmed through the doc too fast). Would you want the metadata in the `main.json` however? The following gets added now:
> 
> image
> 
> ```
> bicep --version
> Bicep CLI version 0.19.5 (87ca110fc0)
> ```
> 
> As for the test env, I currently do not have a functioning one. I can set it up but will be a bit later this week.

Nah, was working on that this week :) 
image

**Comment by MariusStorhaug** [comment]
Good good! Do you want me to remove anything?

Will get the pipeline test this evening if time permits.

**Comment by AlexanderSehr** [comment]
> Good good! Do you want me to remove anything?
> 
> Will get the pipeline test this evening if time permits.

Nope, all good. Will do a final review over the weekend.

**Comment by MariusStorhaug** [comment]
@AlexanderSehr think all is good now. Added the pipeline, had to change params in the encr example as the new param cannot be used with a recoveryPolicy.

File: modules/storage/storage-accounts/.test/encr/main.test.bicep [hunk 1]
```diff
@@ -87,12 +87,10 @@ module testDeployment '../../main.bicep' = {
       containerDeleteRetentionPolicyDays: 10
       containerDeleteRetentionPolicyAllowPermanentDelete: true
       defaultServiceVersion: '2008-10-27'
-      deleteRetentionPolicy: true
+      deleteRetentionPolicyEnabled: true
       deleteRetentionPolicyDays: 9
       isVersioningEnabled: true
       lastAccessTimeTrackingPolicyEnable: true
-      restorePolicyEnabled: true
-      restorePolicyDays: 8
```
Comment (MariusStorhaug) [inline modules/storage/storage-accounts/.test/encr/main.test.bicep:95]: Removed as the default for the new parameter `allowPermanentDelete` is `true`:

https://learn.microsoft.com/en-us/azure/templates/microsoft.storage/storageaccounts/blobservices?pivots=deployment-language-bicep#deleteretentionpolicy

Comment (MariusStorhaug) [reply modules/storage/storage-accounts/.test/encr/main.test.bicep]: testing again with changing the conditions of in the module to be set to `null` if set to false.

File: modules/storage/storage-accounts/blob-services/main.bicep [hunk 2]
```diff
@@ -173,6 +176,7 @@ resource blobServices 'Microsoft.Storage/storageAccounts/blobServices@2022-09-01
     deleteRetentionPolicy: {
       enabled: deleteRetentionPolicyEnabled
       days: deleteRetentionPolicyEnabled == true ? deleteRetentionPolicyDays : null
+      allowPermanentDelete: deleteRetentionPolicyEnabled == true && deleteRetentionPolicyAllowPermanentDelete == true ? deleteRetentionPolicyAllowPermanentDelete : null
```
Comment (MariusStorhaug) [inline modules/storage/storage-accounts/blob-services/main.bicep:179]: When `restorePolicy` is used, `allowPermanentDelete` need to be set to `null`. If provided as `false`, the resource deployment fails with the following error.

https://github.com/MariusStorhaug/ResourceModules/actions/runs/5776225201/job/15655151602#step:4:620

**Comment by MariusStorhaug** [comment]
Ready now.

File: modules/storage/storage-accounts/blob-services/main.bicep (hunk 2 above)
Comment (AlexanderSehr) [inline modules/storage/storage-accounts/blob-services/main.bicep:179]:
```suggestion
      allowPermanentDelete: deleteRetentionPolicyEnabled && deleteRetentionPolicyAllowPermanentDelete ? true : null
```

Would you think I this works too? It's a bit shorter

Comment (MariusStorhaug) [reply modules/storage/storage-accounts/blob-services/main.bicep:179]: It is essentially the same yes...

**Review by AlexanderSehr — APPROVED** [review]

**Pull request merged**